File: backend/backend-widgets/src/main/kotlin/br/com/zup/beagle/widget/action/SendRequest.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.widget.action

import br.com.zup.beagle.widget.context.Bind

data class SendRequest(
    val url: Bind<String>,
    val method: Bind<RequestActionMethod> = Bind.Value(RequestActionMethod.GET),
    val headers: Bind<Map<String, String>>? = null,
    val data: String? = null,
    val onSuccess: Action? = null,
    val onError: Action? = null,
    val onFinish: Action? = null
) : Action {
    constructor(
        url: String,
        method: RequestActionMethod = RequestActionMethod.GET,
        headers: Map<String, String>? = null,
        dataValue: String? = null,
        onSuccess: Action? = null,
        onError: Action? = null,
        onFinish: Action? = null
    ) : this(
        Bind.Value(url),
        Bind.Value(method),
        if (headers != null) Bind.Value(headers) else headers,
        dataValue,
        onSuccess,
        onError,
        onFinish
    )
}

@SuppressWarnings("UNUSED_PARAMETER")
enum class RequestActionMethod {
    GET,
    POST,
    PUT,
    DELETE,
    HEAD,
    PATCH
}

File: schema-ruby-erb/model_generation/Models/Action/send_request.rb
#   Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA

#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
 
#       http://www.apache.org/licenses/LICENSE-2.0
 
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

require_relative '../../Synthax/Attributes/field.rb'
require_relative '../../Synthax/Attributes/dictionary.rb'
require_relative '../base_component.rb'
require_relative '../../Synthax/Types/built_in_type.rb'

class SendRequest < BaseComponent

    def initialize
        action = Action.new
        method = HTTPMethod.new
        variables = [
            Field.new(
                :name => "url",
                :typeName => TypesToString.string,
                :isBindable => true
            ),
            Field.new(
                :name => "method",
                :typeName => method.name,
                :isOptional => true,
                :isBindable => true
            ),
            Field.new(
                :name => "data",
                :typeName => "ContextData",
                :isOptional => true
            ),
            Dictionary.new(
                :name => "headers",
                :type_of_key => TypesToString.string,
                :type_of_value => TypesToString.string,
                :isOptional => true,
                :isBindable => true
            ),
            List.new(
                :name => "onSuccess",
                :typeName => action.name,
                :isOptional => true
            ),
            List.new(
                :name => "onError",
                :typeName => action.name,
                :isOptional => true,
            ),
            List.new(
                :name => "onFinish",
                :typeName => action.name,
                :isOptional => true
            ),


        ]
        synthax_type = BuiltInType.new(
            :name => self.name,
            :variables => variables,
            :package => "br.com.zup.beagle.widget.core",
            :inheritFrom => [action.name],
            :sameFileTypes => [method]
        )

        super(synthax_type)

    end

end

class HTTPMethod < BaseComponent

    def initialize
        enum_cases = [
            EnumCase.new(:name => "get", :defaultValue => "GET"),
            EnumCase.new(:name => "post", :defaultValue => "POST"),
            EnumCase.new(:name => "put", :defaultValue => "PUT"),
            EnumCase.new(:name => "patch", :defaultValue => "PATCH"),
            EnumCase.new(:name => "delete", :defaultValue => "DELETE")
        ]
        synthax_type = EnumType.new(
            :name => self.name,
            :variables => enum_cases,
            :package => "br.com.zup.beagle.widget.core"
        )

        super(synthax_type)

    end

end